lib/gateway_traversal.go

```diff
@@ -0,0 +1,284 @@
+	var err error
+
+	// If the scope is the block, we only need the root block of the last element of the path, which we have.
+	if params.Scope == gateway.DagScopeBlock {
+		return nil
+	}
+
+	// decode the terminal block into a node
+	pc := dagpb.AddSupportToChooser(func(lnk ipld.Link, lnkCtx ipld.LinkContext) (ipld.NodePrototype, error) {
+		if tlnkNd, ok := lnkCtx.LinkNode.(schema.TypedLinkNode); ok {
+			return tlnkNd.LinkTargetNodePrototype(), nil
+		}
+		return basicnode.Prototype.Any, nil
+	})
+
+	pathTerminalCidLink := cidlink.Link{Cid: terminalBlk.Cid()}
+	np, err := pc(pathTerminalCidLink, lctx)
+	if err != nil {
+		return err
+	}
+
+	decoder, err := lsys.DecoderChooser(pathTerminalCidLink)
+	if err != nil {
+		return err
+	}
+	nb := np.NewBuilder()
+	blockData := terminalBlk.RawData()
+	if err := decoder(nb, bytes.NewReader(blockData)); err != nil {
+		return err
+	}
+	lastCidNode := nb.Build()
+
+	// TODO: Evaluate:
+	// Does it matter that we're ignoring the "remainder" portion of the traversal in GetCAR?
+	// Does it matter that we're using a linksystem with the UnixFS reifier for dagscope=all?
+
+	// If we're asking for everything then give it
+	if params.Scope == gateway.DagScopeAll {
+		sel, err := selector.ParseSelector(selectorparse.CommonSelector_ExploreAllRecursively)
+		if err != nil {
+			return err
+		}
+
+		progress := traversal.Progress{
+			Cfg: &traversal.Config{
+				Ctx:                            ctx,
+				LinkSystem:                     *lsys,
+				LinkTargetNodePrototypeChooser: bsfetcher.DefaultPrototypeChooser,
+				LinkVisitOnlyOnce:              true, // This is safe for the "all" selector
+			},
+		}
+
+		if err := progress.WalkMatching(lastCidNode, sel, func(progress traversal.Progress, node datamodel.Node) error {
+			return nil
+		}); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	// From now on, dag-scope=entity!
+	// Since we need more of the graph load it to figure out what we have
+	// This includes determining if the terminal node is UnixFS or not
+	if pbn, ok := lastCidNode.(dagpb.PBNode); !ok {
+		// If it's not valid dag-pb then we're done
+		return nil
+	} else if !pbn.FieldData().Exists() {
+		// If it's not valid UnixFS then we're done
+		return nil
+	} else if unixfsFieldData, decodeErr := data.DecodeUnixFSData(pbn.Data.Must().Bytes()); decodeErr != nil {
+		// If it's not valid dag-pb and UnixFS then we're done
+		return nil
+	} else {
+		switch unixfsFieldData.FieldDataType().Int() {
+		case data.Data_Directory, data.Data_Symlink:
+			// These types are non-recursive so we're done
+			return nil
+		case data.Data_Raw, data.Data_Metadata:
+			// TODO: for now, we decided to return nil here. The different implementations are inconsistent
+			// and UnixFS is not properly specified: https://github.com/ipfs/specs/issues/316.
+			// 		- Is Data_Raw different from Data_File?
+			//		- Data_Metadata is handled differently in boxo/ipld/unixfs and go-unixfsnode.
+			return nil
```

Comment:
:thought_balloon:  do we know what was the old behavior in Kubo?
Unsure how popular these are, but if we don't have spec, we may want to derisk and not change behavior that has been in production for years.

Reply:
We sort of ended up giving this a 🤷 anyhow here https://github.com/ipfs/boxo/pull/303#discussion_r1195267844. We can revisit both, but don't think that needs to happen in this PR.